File: Application/src/main/java/com/shareyourproxy/app/dialog/AddChannelDialog.java
package com.shareyourproxy.app.dialog;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialog;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;

import com.shareyourproxy.R;
import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.ChannelType;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxBusDriver;
import com.shareyourproxy.api.rx.command.AddUserChannelCommand;
import com.shareyourproxy.api.rx.event.AddChannelDialogSuccessEvent;

import java.util.UUID;

import butterknife.Bind;
import butterknife.BindColor;
import butterknife.BindString;
import butterknife.ButterKnife;

import static android.content.DialogInterface.OnClickListener;
import static com.shareyourproxy.api.domain.factory.ChannelFactory.createModelInstance;
import static com.shareyourproxy.util.ObjectUtils.getSimpleName;
import static com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard;

/**
 * Add a new {@link Channel} to a {@link User}.
 */
public class AddChannelDialog extends BaseDialogFragment {
    private static final String ARG_CHANNEL_TYPE = "AddChannelDialog.ChannelType";
    private static final String TAG = Companion.getSimpleName(AddChannelDialog.class);
    @Bind(R.id.dialog_channel_action_address_edittext)
    EditText editTextActionAddress;
    private final OnClickListener _negativeClicked =
        new OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                hideSoftwareKeyboard(editTextActionAddress);
                dismiss();
            }
        };
    @Bind(R.id.dialog_channel_label_edittext)
    EditText editTextLabel;
    @Bind(R.id.dialog_channel_label_floatlabel)
    TextInputLayout floatLabelChannelLabel;
    @Bind(R.id.dialog_channel_action_address_floatlabel)
    TextInputLayout floatLabelAddress;
    @BindColor(R.color.common_text)
    int colorText;
    @BindColor(R.color.common_divider)
    int colorGray;
    @BindColor(R.color.common_blue)
    int colorBlue;
    @BindString(R.string.required)
    String _required;
    private ChannelType _channelType;
    /**
     * EditorActionListener that detects when the software keyboard's done or enter button is pressed.
     */
    private final OnEditorActionListener _onEditorActionListener =
        new OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                // KeyEvent.KEYCODE_ENDCALL is the actionID of the Done button when this
                // FixedDecimalEditText's inputType is Decimal
                if (actionId == KeyEvent.KEYCODE_ENTER
                    || actionId == KeyEvent.KEYCODE_ENDCALL) {
                    saveChannelAndExit();
                    return true;
                }
                return false;
            }
        };
    private final View.OnClickListener _positiveClicked =
        new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveChannelAndExit();
            }
        };
    private String _dialogTitle;
    private String _channelAddressHint;
    private String _channelLabelHint;

    /**
     * Constructor.
     */
    public AddChannelDialog() {
    }

    /**
     * Create a new instance of a {@link AddChannelDialog}.
     *
     * @return A {@link AddChannelDialog}
     */
    public static AddChannelDialog newInstance(ChannelType channelType) {
        //Bundle arguments
        Bundle bundle = new Bundle();
        bundle.putString(ARG_CHANNEL_TYPE, channelType.getLabel());
        //create dialog instance
        AddChannelDialog dialog = new AddChannelDialog();
        dialog.setArguments(bundle);
        return dialog;
    }

    /**
     * Dispatch a Channel Added Event
     */
    private void addUserChannel() {
        //TODO:CLEAN THIS MESS UP
        String actionContent = editTextActionAddress.getText().toString();
        String labelContent = editTextLabel.getText().toString().trim();
        if (!TextUtils.isEmpty(actionContent.trim())) {
            String id = UUID.randomUUID().toString();
            Channel channel =
                createModelInstance(id, labelContent, _channelType, actionContent);
            RxBusDriver rxBus = getRxBus();
            User user = getLoggedInUser();
            rxBus.post(new AddUserChannelCommand(user, channel));
            user.channels().put(channel.id(), channel);
            rxBus.post(new AddChannelDialogSuccessEvent(user, channel));
        }
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        _channelType = ChannelType.Companion.valueOfLabel(getArguments().getString(ARG_CHANNEL_TYPE));
    }

    @NonNull
    @Override
    @SuppressLint("InflateParams")
    public AppCompatDialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreateDialog(savedInstanceState);
        View view = getActivity().getLayoutInflater()
            .inflate(R.layout.dialog_add_channel, null, false);
        ButterKnife.bind(this, view);
        initializeDisplayValues();

        editTextActionAddress.setOnEditorActionListener(_onEditorActionListener);
        AlertDialog dialog = new AlertDialog.Builder(getActivity(),
            R.style.Widget_Proxy_App_Dialog)
            .setTitle(_dialogTitle)
            .setView(view)
            .setPositiveButton(getString(R.string.save), null)
            .setNegativeButton(android.R.string.cancel, _negativeClicked)
            .create();

        // Show the SW Keyboard on dialog start. Always.
        dialog.getWindow().setSoftInputMode(
            WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        dialog.getWindow().getAttributes().width = WindowManager.LayoutParams.MATCH_PARENT;
        dialog.setCanceledOnTouchOutside(false);
        return dialog;
    }

    @Override
    public void onStart() {
        super.onStart();
        AlertDialog dialog = (AlertDialog) getDialog();
        setButtonTint(dialog.getButton(Dialog.BUTTON_POSITIVE), colorBlue);
        setButtonTint(dialog.getButton(Dialog.BUTTON_NEGATIVE), colorText);
        //Alert Dialogs dismiss by default because of an internal handler... this bypasses that.
        dialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(_positiveClicked);
        //Setup TextInput hints.
        floatLabelAddress.setHint(_channelAddressHint);
        floatLabelChannelLabel.setHint(_channelLabelHint);
    }

    public void saveChannelAndExit() {
        boolean addressHasText = editTextActionAddress.getText().toString().trim().length() > 0;
        if (!addressHasText) {
            floatLabelAddress.setError(_required);
        } else {
            floatLabelAddress.setErrorEnabled(false);
            addUserChannel();
            dismiss();
        }
    }

    private void initializeDisplayValues() {
        String name = _channelType.getLabel();
        switch (_channelType) {
            case Address:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Custom:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_custom);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_custom);
                break;
            case Ello:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Email:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_email);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_email);
                break;
            case Facebook:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_default);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_default);
            case FBMessenger:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Github:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Googleplus:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Hangouts:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Instagram:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case LeagueOfLegends:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.username);
                _channelLabelHint = getString(R.string.label);
                break;
            case Linkedin:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Medium:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Meerkat:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case NintendoNetwork:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.username);
                _channelLabelHint = getString(R.string.label);
                break;
            case Periscope:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Phone:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_channel_hint_address_phone);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_phone);
                break;
            case PlaystationNetwork:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.username);
                _channelLabelHint = getString(R.string.label);
                break;
            case Reddit:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_default);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_default);
            case Skype:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Slack:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case SMS:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_sms);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_sms);
                break;
            case Snapchat:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Soundcloud:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Spotify:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Steam:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.username);
                _channelLabelHint = getString(R.string.label);
                break;
            case Tumblr:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Twitch:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.username);
                _channelLabelHint = getString(R.string.label);
                break;
            case Twitter:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_default);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_default);
            case Venmo:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Web:
            case URL:
                _dialogTitle = getString(R.string.dialog_addchannel_title_web);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_web);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_web);
                break;
            case Whatsapp:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_phone);
                _channelLabelHint = getString(R.string.label);
                break;
            case XboxLive:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(R.string.gamertag);
                _channelLabelHint = getString(R.string.label);
                break;
            case Yo:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            case Youtube:
                _dialogTitle = getString(R.string.dialog_addchannel_title_add_blank, name);
                _channelAddressHint = getString(
                    R.string.dialog_channel_hint_address_blank_handle, name);
                _channelLabelHint = getString(R.string.label);
                break;
            default:
                _dialogTitle = getString(R.string.dialog_addchannel_title_default);
                _channelAddressHint = getString(R.string.dialog_addchannel_hint_address_default);
                _channelLabelHint = getString(R.string.dialog_addchannel_hint_label_default);
                break;
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    /**
     * Use the private string TAG from this class as an identifier.
     *
     * @param fragmentManager manager of fragments
     * @return this dialog
     */
    public AddChannelDialog show(FragmentManager fragmentManager) {
        show(fragmentManager, TAG);
        return this;
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/app/dialog/AddChannelDialog.kt
package com.shareyourproxy.app.dialog

import android.R.string.cancel
import android.app.Dialog
import android.content.DialogInterface
import android.content.DialogInterface.OnClickListener
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatDialog
import android.text.TextUtils
import android.view.KeyEvent.KEYCODE_ENDCALL
import android.view.KeyEvent.KEYCODE_ENTER
import android.view.View
import android.view.WindowManager.LayoutParams.MATCH_PARENT
import android.view.WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE
import android.widget.EditText
import android.widget.TextView.OnEditorActionListener
import com.shareyourproxy.R
import com.shareyourproxy.R.color.common_blue
import com.shareyourproxy.R.color.common_text
import com.shareyourproxy.R.id.*
import com.shareyourproxy.R.layout.dialog_add_channel
import com.shareyourproxy.R.string.*
import com.shareyourproxy.R.style.Widget_Proxy_App_Dialog
import com.shareyourproxy.api.domain.factory.ChannelFactory.createModelInstance
import com.shareyourproxy.api.domain.model.ChannelType
import com.shareyourproxy.api.domain.model.ChannelType.*
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.command.AddUserChannelCommand
import com.shareyourproxy.api.rx.event.AddChannelDialogSuccessEvent
import com.shareyourproxy.util.ButterKnife.LazyVal
import com.shareyourproxy.util.ButterKnife.bindColor
import com.shareyourproxy.util.ButterKnife.bindString
import com.shareyourproxy.util.ButterKnife.bindView
import com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard
import java.util.*

/**
 * Add a new [Channel] to a [User].
 */
internal final class AddChannelDialog private constructor(channelType: ChannelType) : BaseDialogFragment() {
    companion object {
        private val ARG_CHANNEL_TYPE = "AddChannelDialog.ChannelType"
        fun show(manager: FragmentManager, channelType: ChannelType): AddChannelDialog {
            return setArgs(manager, channelType)
        }

        private fun setArgs(manager: FragmentManager, channelType: ChannelType): AddChannelDialog {
            val dialog = AddChannelDialog(channelType)
            val args: Bundle = Bundle()
            args.putSerializable(ARG_CHANNEL_TYPE, channelType)
            dialog.arguments = args
            return dialog.show(manager)
        }
    }

    private val editTextActionAddress: EditText by bindView(dialog_channel_action_address_edittext)
    private val negativeClicked = OnClickListener { dialogInterface, i ->
        hideSoftwareKeyboard(editTextActionAddress)
        dismiss()
    }
    private val editTextLabel: EditText by bindView(dialog_channel_label_edittext)
    private val floatLabelChannelLabel: TextInputLayout by bindView(dialog_channel_label_floatlabel)
    private val floatLabelAddress: TextInputLayout by bindView(dialog_channel_action_address_floatlabel)
    private val colorText: Int by bindColor(common_text)
    private val colorBlue: Int by bindColor(common_blue)
    private val stringRequired: String by bindString(required)
    private val parcelChannelType: ChannelType by LazyVal{ arguments.getSerializable(ARG_CHANNEL_TYPE) as ChannelType}
    /**
     * EditorActionListener that detects when the software keyboard's done or enter button is pressed.
     */
    private val onEditorActionListener = OnEditorActionListener { v, actionId, event ->
        // KeyEvent.KEYCODE_ENDCALL is the actionID of the Done button
        when(actionId){
            KEYCODE_ENDCALL,
            KEYCODE_ENTER -> saveChannelAndExit()
            else -> false
        }
    }
    private val positiveClicked = View.OnClickListener { saveChannelAndExit() }
    private var dialogTitle: String = ""
    private var channelAddressHint: String = ""
    private var channelLabelHint: String = ""

    override fun onCreateDialog(savedInstanceState: Bundle?): AppCompatDialog {
        super.onCreateDialog(savedInstanceState)
        val view = activity.layoutInflater.inflate(dialog_add_channel, null, false)
        initializeDisplayValues()

        editTextActionAddress.setOnEditorActionListener(onEditorActionListener)
        val dialog = AlertDialog.Builder(activity, Widget_Proxy_App_Dialog)
                .setTitle(dialogTitle)
                .setView(view)
                .setPositiveButton(save, null)
                .setNegativeButton(cancel, negativeClicked)
                .create()

        // Show the SW Keyboard on dialog start. Always.
        dialog.window.setSoftInputMode(SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        dialog.window.attributes.width = MATCH_PARENT
        dialog.setCanceledOnTouchOutside(false)
        return dialog
    }

    override fun onStart() {
        super.onStart()
        val dialog = dialog as AlertDialog
        setButtonTint(dialog.getButton(Dialog.BUTTON_POSITIVE), colorBlue)
        setButtonTint(dialog.getButton(Dialog.BUTTON_NEGATIVE), colorText)
        //Alert Dialogs dismiss by default because of an internal handler... this bypasses that.
        dialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(positiveClicked)
        //Setup TextInput hints.
        floatLabelAddress.hint = channelAddressHint
        floatLabelChannelLabel.hint = channelLabelHint
    }

    private fun saveChannelAndExit(): Boolean {
        val addressHasText = editTextActionAddress.text.toString().trim { it <= ' ' }.length > 0
        if (!addressHasText) {
            floatLabelAddress.error = stringRequired
        } else {
            floatLabelAddress.isErrorEnabled = false
            addUserChannel()
            dismiss()
        }
        return true
    }

    private fun initializeDisplayValues() {
        val name = parcelChannelType.label
        when (parcelChannelType) {
            Address -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Custom -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_addchannel_hint_address_custom)
                channelLabelHint = getString(dialog_addchannel_hint_label_custom)
            }
            Ello -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Email -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_addchannel_hint_address_email)
                channelLabelHint = getString(dialog_addchannel_hint_label_email)
            }
            Facebook -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            FBMessenger -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Github -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Googleplus -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Hangouts -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Instagram -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            LeagueOfLegends -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(username)
                channelLabelHint = getString(label)
            }
            Linkedin -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Medium -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Meerkat -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            NintendoNetwork -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(username)
                channelLabelHint = getString(label)
            }
            Periscope -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Phone -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_phone)
                channelLabelHint = getString(dialog_addchannel_hint_label_phone)
            }
            PlaystationNetwork -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(username)
                channelLabelHint = getString(label)
            }
            Reddit -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Skype -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Slack -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            SMS -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_addchannel_hint_address_sms)
                channelLabelHint = getString(dialog_addchannel_hint_label_sms)
            }
            Snapchat -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Soundcloud -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Spotify -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Steam -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(username)
                channelLabelHint = getString(label)
            }
            Tumblr -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Twitch -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(username)
                channelLabelHint = getString(label)
            }
            Twitter -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_addchannel_hint_address_default)
                channelLabelHint = getString(dialog_addchannel_hint_label_default)
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Venmo -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Web, URL -> {
                dialogTitle = getString(dialog_addchannel_title_web)
                channelAddressHint = getString(dialog_addchannel_hint_address_web)
                channelLabelHint = getString(dialog_addchannel_hint_label_web)
            }
            Whatsapp -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_phone)
                channelLabelHint = getString(label)
            }
            XboxLive -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(gamertag)
                channelLabelHint = getString(label)
            }
            Yo -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            Youtube -> {
                dialogTitle = getString(dialog_addchannel_title_add_blank, name)
                channelAddressHint = getString(dialog_channel_hint_address_blank_handle, name)
                channelLabelHint = getString(label)
            }
            else -> {
                dialogTitle = getString(R.string.dialog_addchannel_title_default)
                channelAddressHint = getString(dialog_addchannel_hint_address_default)
                channelLabelHint = getString(dialog_addchannel_hint_label_default)
            }
        }
    }

    /**
     * Dispatch a Channel Added Event
     */
    private fun addUserChannel() {
        val actionContent = editTextActionAddress.text.toString()
        val labelContent = editTextLabel.text.toString().trim { it <= ' ' }
        if (!TextUtils.isEmpty(actionContent.trim { it <= ' ' })) {
            val id = UUID.randomUUID().toString()
            val channel = createModelInstance(id, labelContent, parcelChannelType, actionContent)
            val user = loggedInUser
            post(AddUserChannelCommand(user, channel))
            user.channels.put(channel.id, channel)
            post(AddChannelDialogSuccessEvent(user, channel))
        }
    }

    /**
     * Use the private string TAG from this class as an identifier.
     * @param fragmentManager manager of fragments
     * @return this dialog
     */
    fun show(fragmentManager: FragmentManager): AddChannelDialog {
        show(fragmentManager, AddChannelDialog::class.java.simpleName)
        return this
    }
}
